### PR TITLE
Rewrite BTC-ETH e2e happy path test

### LINKED ISSUE #1246: Rewrite the first happy path e2e test (bitcoin <> ether)
[Spike 11](https://github.com/comit-network/spikes/blob/master/0011-revise-e2e-test-framework.adoc) recommends a new e2e test design.
This ticket is about changing an initial test to this new design.

We recommend to start with "bitcoin" <> "ether".
Get inspired by the suggested design here: https://github.com/comit-network/spikes/blob/master/0011-revise-e2e-test-framework.adoc#the-ideal-test

This ticket on purpose doesn't include anything about the test-harness. You are only supposed to implement the test "framework".

Use the sdk.

### PR DESCRIPTION
**Pull request by luckysori**
The bitcoin for ether happy path test now looks like what was proposed in [spike 11](https://github.com/comit-network/spikes/blob/master/0011-revise-e2e-test-framework.adoc#the-ideal-test).

Notes:
- This introduces a secondary library `lib2` inside `api_tests` which will eventually replace `lib`.
- My work builds on top of @thomaseizinger's [branch](https://github.com/comit-network/comit-rs/tree/the-perfect-test) from a while ago.
- Currently blocked by some [minor changes to the SDK](https://github.com/comit-network/comit-js-sdk/pull/26).
## The more ideal test

```rust
import { createActors } from "../../../lib2/create_actors";

setTimeout(function() {
    describe("happy path", function() {
        this.timeout(20000);
        it("bitcoin ether", async function() {
            const { alice, bob } = await createActors(
                "e2e-rfc003-btc-eth-happy.log"
            );

            await alice.sendRequest("bitcoin", "ether");
            await bob.accept();
            await alice.fund();
            await bob.fund();
            await alice.redeem();
            await bob.redeem();

            await alice.assertSwapped();
            await bob.assertSwapped();
        });
    });
    run();
}, 0);

```

## Command line

Oct28::184449

## Logs

Located in file  `api_tests\log\tests\e2e-rfc003-btc-eth-happy.log`.

```
[2019-10-28T18:48:56.342] [INFO] alice - Created new actor at http://localhost:8000
[2019-10-28T18:48:56.344] [INFO] bob - Created new actor at http://localhost:8010
[2019-10-28T18:48:56.344] [INFO] alice - Sending swap request
[2019-10-28T18:48:56.344] [DEBUG] alice - Derived { name: 'bitcoin', network: 'regtest' } from asset bitcoin
[2019-10-28T18:48:56.345] [DEBUG] alice - Derived { name: 'bitcoin', quantity: '100000000' } from asset bitcoin
[2019-10-28T18:48:56.345] [DEBUG] alice - Derived { name: 'ethereum', network: 'regtest' } from asset ether
[2019-10-28T18:48:56.345] [DEBUG] alice - Derived { name: 'ether', quantity: '10000000000000000000' } from asset ether
[2019-10-28T18:48:56.711] [DEBUG] alice - Minting bitcoin on bitcoin
[2019-10-28T18:48:57.969] [DEBUG] alice - Starting bitcoin balance:  200000000
[2019-10-28T18:48:57.970] [DEBUG] bob - Minting ether on ethereum
[2019-10-28T18:48:58.511] [DEBUG] bob - Starting ether balance:  20000000000000000000
[2019-10-28T18:48:58.642] [DEBUG] alice - Created new swap at /swaps/rfc003/f2ddd199-8be0-4bc0-9010-3deb62ed1c4d
[2019-10-28T18:48:58.643] [DEBUG] alice - Funding as part of swap @ /swaps/rfc003/f2ddd199-8be0-4bc0-9010-3deb62ed1c4d
[2019-10-28T18:49:02.648] [DEBUG] bob - Funding as part of swap @ /swaps/rfc003/f2ddd199-8be0-4bc0-9010-3deb62ed1c4d
[2019-10-28T18:49:02.803] [DEBUG] alice - Redeeming as part of swap @ /swaps/rfc003/f2ddd199-8be0-4bc0-9010-3deb62ed1c4d
[2019-10-28T18:49:03.957] [DEBUG] bob - Redeeming as part of swap @ /swaps/rfc003/f2ddd199-8be0-4bc0-9010-3deb62ed1c4d
[2019-10-28T18:49:08.718] [DEBUG] alice - Checking if swap @ /swaps/rfc003/f2ddd199-8be0-4bc0-9010-3deb62ed1c4d is done
[2019-10-28T18:49:08.718] [DEBUG] alice - Checking that ether balance changed by 10000000000000000000
[2019-10-28T18:49:09.730] [DEBUG] bob - Checking if swap @ /swaps/rfc003/f2ddd199-8be0-4bc0-9010-3deb62ed1c4d is done
[2019-10-28T18:49:09.730] [DEBUG] bob - Checking that bitcoin balance changed by 100000000
```

---

Fixes #1246.

### CODE REVIEW

File: api_tests/lib2/actors/actor.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        await this.setStartingBalance([
+            alphaAsset,
+            { name: betaAsset.name, quantity: 0 },
+        ]);
+        await to.setStartingBalance([
+            { name: alphaAsset.name, quantity: 0 },
+            betaAsset,
+        ]);
+
+        // this.expectedBalanceChanges.set(alphaAssetKind, -alphaAsset.quantity);
```
Comment (D4nte) [inline api_tests/lib2/actors/actor.ts]:
```suggestion
```

File: api_tests/lib2/actors/actor.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        this.expectedBalanceChanges.set(betaAssetKind, betaAsset.quantity);
+        to.expectedBalanceChanges.set(alphaAssetKind, alphaAsset.quantity);
+        // to.expectedBalanceChanges.set(betaAssetKind, -betaAsset.quantity);
```
Comment (D4nte) [inline api_tests/lib2/actors/actor.ts]:
```suggestion
```

File: api_tests/lib2/actors/actor.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+            to.wallets.bitcoin.inner,
+            to.wallets.ethereum.inner,
+            to.cnd,
+            this.swap.self
+        );
+        this.logger.debug("Created new swap at %s", this.swap.self);
+    }
+
+    public async accept() {
+        this.swap.accept(Actor.defaultActionConfig);
```
Comment (D4nte) [inline api_tests/lib2/actors/actor.ts]: You check the swap existence for funding but not accepting.

File: api_tests/lib2/actors/actor.ts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+            const actualBalance = await wallet.getBalance();
+            const expectedBalance =
+                this.startingBalances.get(assetKind) + expectedBalanceChange;
+            const maximumFee = wallet.MaximumFee;
+
+            expect(actualBalance).to.be.at.least(expectedBalance - maximumFee);
+        }
+    }
+
+    private async additionalIdentities(
```
Comment (D4nte) [inline api_tests/lib2/actors/actor.ts]: This is definitely something that should be part of the sdk or API should be improved :(

File: api_tests/lib2/wallets/index.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+declare var global: HarnessGlobal;
+
+interface AllWallets {
+    bitcoin?: BitcoinWallet;
+    ethereum?: EthereumWallet;
+}
+
+export type LedgerKind = keyof AllWallets;
+export type AssetKind = "bitcoin" | "ether" | "erc20";
```
Comment (D4nte) [inline api_tests/lib2/wallets/index.ts]:
```suggestion
export enum AssetKind {
    Bitcoin,
    Ether,
    Erc20
}
```
Why not using an enum?

File: api_tests/package.json [hunk 6]
```diff
@@ -23,8 +24,9 @@
         "chai-http": "^4.2.0",
         "chai-json-schema": "^1.5.0",
         "commander": "^3.0.2",
-        "ethers": "^4.0.38",
```
Comment (D4nte) [inline api_tests/package.json]: Should we update ethers in sdk instead?

File: api_tests/lib2/actors/actor.ts [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+                name: "ethereum",
+                network: "regtest",
+            };
+        }
+    }
+}
+
+function defaultAssetDescriptionForAsset(asset: AssetKind): Asset {
+    switch (asset) {
+        case "bitcoin": {
```
Comment (D4nte) [inline api_tests/lib2/actors/actor.ts]:
```suggestion
        case AssetKind.Bitcoin: {
```
Just a preference

File: api_tests/e2e/rfc003/btc_eth/happy.ts [hunk 8]
```diff
@@ -1,123 +1,23 @@
...
-                },
-            },
-        },
-    ];
-
-    describe("RFC003: Bitcoin for Ether", async () => {
-        createTests(alice, bob, steps, "/swaps/rfc003", "/swaps", swapRequest);
+import { createActors } from "../../../lib2/create_actors";
+
+setTimeout(function() {
```
[the remark this answers is not in the document]
Comment (luckysori) [reply api_tests/e2e/rfc003/btc_eth/happy.ts]: Sadly it appears to be necessary fort the test to run. I believe it is some sort of simple hack that @thomaseizinger came up with to make it work. I also don't understand why it is needed for this new sort of test, but not for the old ones.

[the remark this answers is not in the document]
Comment (D4nte) [reply api_tests/e2e/rfc003/btc_eth/happy.ts]: yeah, actually it was my hack :) just wanted to double check. It's because we use async mocha.

**Pull request merged**